File: py_version/solve_text_txt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
BroLang Decoder — Python version (file I/O)
Dùng: python solve_text.py <input_file> <output_file>
"""

import sys
import re

# ──────────────────────────────────────────────
# Bảng mã
# ──────────────────────────────────────────────

# s1–s26: a–z
BASE_MAP = {f"s{i}": chr(ord('a') + i - 1) for i in range(1, 27)}
# s27: ,  s28: .
BASE_MAP["s27"] = ","
BASE_MAP["s28"] = "."
# s0: space
BASE_MAP["s0"] = " "
# c0–c9: chữ số
for _i in range(10):
    BASE_MAP[f"c{_i}"] = str(_i)

# Dấu phụ: base_vowel → { modifier → char }
DIACRITIC_MAP = {
    'a': {"a1": "ă", "a2": "â"},
    'e': {"a2": "ê"},
    'o': {"a2": "ô", "a3": "ơ"},
    'u': {"a3": "ư"},
}

# Thanh điệu: base_char → { modifier → char }
TONE_MAP = {}

def _add_tones(base, sac, huyen, hoi, nga, nang):
    TONE_MAP[base] = {"a4": sac, "a5": huyen, "a6": hoi, "a7": nga, "a8": nang}

_add_tones("a","á","à","ả","ã","ạ"); _add_tones("A","Á","À","Ả","Ã","Ạ")
_add_tones("ă","ắ","ằ","ẳ","ẵ","ặ"); _add_tones("Ă","Ắ","Ằ","Ẳ","Ẵ","Ặ")
_add_tones("â","ấ","ầ","ẩ","ẫ","ậ"); _add_tones("Â","Ấ","Ầ","Ẩ","Ẫ","Ậ")
_add_tones("e","é","è","ẻ","ẽ","ẹ"); _add_tones("E","É","È","Ẻ","Ẽ","Ẹ")
_add_tones("ê","ế","ề","ể","ễ","ệ"); _add_tones("Ê","Ế","Ề","Ể","Ễ","Ệ")
_add_tones("i","í","ì","ỉ","ĩ","ị"); _add_tones("I","Í","Ì","Ỉ","Ĩ","Ị")
_add_tones("o","ó","ò","ỏ","õ","ọ"); _add_tones("O","Ó","Ò","Ỏ","Õ","Ọ")
_add_tones("ô","ố","ồ","ổ","ỗ","ộ"); _add_tones("Ô","Ố","Ồ","Ổ","Ỗ","Ộ")
_add_tones("ơ","ớ","ờ","ở","ỡ","ợ"); _add_tones("Ơ","Ớ","Ờ","Ở","Ỡ","Ợ")
_add_tones("u","ú","ù","ủ","ũ","ụ"); _add_tones("U","Ú","Ù","Ủ","Ũ","Ụ")
_add_tones("ư","ứ","ừ","ử","ữ","ự"); _add_tones("Ư","Ứ","Ừ","Ử","Ữ","Ự")
_add_tones("y","ý","ỳ","ỷ","ỹ","ỵ"); _add_tones("Y","Ý","Ỳ","Ỷ","Ỹ","Ỵ")

# Dấu phụ viết hoa
UPPER_DIAC = {"ă":"Ă","â":"Â","ê":"Ê","ô":"Ô","ơ":"Ơ","ư":"Ư"}

# Toán tử: t0–t7
T_MAP = {
    "t0": "=", "t1": "+", "t2": "-", "t3": "*",
    "t4": "/", "t5": "^", "t6": "//", "t7": "%",
}


# ──────────────────────────────────────────────
# Decode escape sequence trong oth[...]
# ──────────────────────────────────────────────

def decode_escape(s: str) -> str:
    result = []
    i = 0
    while i < len(s):
        if s[i] == '*' and i + 1 < len(s):
            # *tab → \t
            if s[i+1:i+4] == "tab":
                result.append('\t')
                i += 4
            # *\n → \n
            elif s[i+1:i+3] == "\\n":
                result.append('\n')
                i += 3
            elif s[i+1] == 'n':
                result.append('\n')
                i += 2
            elif s[i+1] == 't':
                result.append('\t')
                i += 2
            else:
                result.append(s[i+1])
                i += 2
        elif s[i] == '\\' and i + 1 < len(s):
            if s[i+1] == 'n':
                result.append('\n'); i += 2
            elif s[i+1] == 't':
                result.append('\t'); i += 2
            else:
                result.append('\\'); result.append(s[i+1]); i += 2
        else:
            result.append(s[i]); i += 1
    return ''.join(result)


# ──────────────────────────────────────────────
# Giải mã một token sX/cX[modifiers]
# ──────────────────────────────────────────────

def solve_token(token: str) -> str:
    bracket = token.find('[')
    base_code = token[:bracket] if bracket != -1 else token

    if base_code not in BASE_MAP:
        return ""
    decoded = BASE_MAP[base_code]

    if bracket == -1:
        return decoded

    mods_str = token[bracket+1:-1]  # nội dung trong []
    mods = mods_str.split()

    make_upper = False
    diacritic_mod = ""
    tone_mod = ""
    repeat = 1

    for m in mods:
        if m == "cl":
            make_upper = True
        elif m == "a0":
            if base_code == "s4":
                return "Đ" if make_upper else "đ"
        elif m in ("a1", "a2", "a3"):
            diacritic_mod = m
        elif m in ("a4", "a5", "a6", "a7", "a8"):
            tone_mod = m
        elif m.isdigit():
            repeat = int(m)

    # a0 có thể đến sau cl → kiểm tra lại
    if "a0" in mods and base_code == "s4":
        return "Đ" if make_upper else "đ"

    # Viết hoa ký tự đơn ASCII
    if make_upper and len(decoded) == 1:
        decoded = decoded.upper()

    # Dấu phụ
    if diacritic_mod and len(decoded) == 1:
        base_vowel = decoded.lower()
        if base_vowel in DIACRITIC_MAP and diacritic_mod in DIACRITIC_MAP[base_vowel]:
            decoded = DIACRITIC_MAP[base_vowel][diacritic_mod]
            if make_upper and decoded in UPPER_DIAC:
                decoded = UPPER_DIAC[decoded]

    # Thanh điệu
    if tone_mod and decoded in TONE_MAP and tone_mod in TONE_MAP[decoded]:
        decoded = TONE_MAP[decoded][tone_mod]

    return decoded * repeat


# ──────────────────────────────────────────────
# Tokenizer
# ──────────────────────────────────────────────

def tokenize(s: str) -> list:
    tokens = []
    current = []
    in_bracket = False

    for ch in s:
        if ch == '[':
            in_bracket = True
            current.append(ch)
        elif ch == ']':
            in_bracket = False
            current.append(ch)
            tokens.append(''.join(current))
            current = []
        elif ch == '(' and not in_bracket:
            pass  # grouping — bỏ qua
        elif ch == ')' and not in_bracket:
            if current:
                tokens.append(''.join(current))
                current = []
        elif ch in (' ', ',', '\t') and not in_bracket:
            if current:
                tokens.append(''.join(current))
                current = []
        else:
            current.append(ch)

    if current:
        tokens.append(''.join(current))
    return tokens


# ──────────────────────────────────────────────
# Decode block nội dung S(...)E
# ──────────────────────────────────────────────

def decode_block(content: str) -> str:
    tokens = tokenize(content)
    result = []

    for token in tokens:
        if token == "s0":
            result.append(" ")
        elif token.startswith("oth[") and token.endswith("]"):
            inner = token[4:-1]
            result.append(decode_escape(inner))
        elif token and token[0] == 'c' and '[' not in token:
            if len(token) > 1 and token[1:].isdigit():
                result.append(token[1:])
        elif token and token[0] in ('s', 'c'):
            result.append(solve_token(token))
        elif token and token[0] == 't' and len(token) > 1 and token[1].isdigit():
            br = token.find('[')
            t_code = token[:br] if br != -1 else token
            if t_code == "t8":
                if br != -1 and token.endswith(']'):
                    inner = token[br+1:-1]
                    result.append("(" + decode_block(inner) + ")")
            elif t_code in T_MAP:
                result.append(T_MAP[t_code])

    return ''.join(result)


# ──────────────────────────────────────────────
# Xử lý một dòng thô (fallback, không có S(...)E)
# ──────────────────────────────────────────────

def process_raw_line(line: str) -> str:
    if not line:
        return ""
    tokens = tokenize(line)
    result = []
    for token in tokens:
        if token == "s0":
            result.append(" ")
        elif token.startswith("oth[") and token.endswith("]"):
            result.append(decode_escape(token[4:-1]))
        elif token and token[0] == 'c' and '[' not in token:
            if len(token) > 1 and token[1:].isdigit():
                result.append(token[1:])
        elif token and token[0] in ('s', 'c'):
            result.append(solve_token(token))
    return ''.join(result)


# ──────────────────────────────────────────────
# Main: đọc file, xử lý S(...)E multiline
# ──────────────────────────────────────────────

def process(input_path: str, output_path: str):
    with open(input_path, 'r', encoding='utf-8') as f:
        full_input = f.read()

    out_parts = []
    pos = 0

    while pos <= len(full_input):
        se_start = full_input.find("S(", pos)

        # Phần thô trước S(
        raw_before = full_input[pos:se_start] if se_start != -1 else full_input[pos:]
        if raw_before:
            for raw_line in raw_before.splitlines(keepends=True):
                stripped = raw_line.rstrip('\n').rstrip('\r')
                out_parts.append(process_raw_line(stripped) + "\n")

        if se_start == -1:
            break

        # Tìm )E đóng — đếm depth, bỏ qua nội dung trong []
        depth = 0
        in_sq = False
        end_pos = -1
        i = se_start + 2
        while i < len(full_input):
            c = full_input[i]
            if c == '[':
                in_sq = True
            elif c == ']':
                in_sq = False
            elif not in_sq:
                if c == '(':
                    depth += 1
                elif c == ')':
                    if depth == 0:
                        if i + 1 < len(full_input) and full_input[i+1] == 'E':
                            end_pos = i
                            break
                    else:
                        depth -= 1
            i += 1

        if end_pos == -1:
            break

        # Trích nội dung, thay newline → space
        inner = full_input[se_start+2:end_pos].replace('\n', ' ')
        out_parts.append(decode_block(inner))

        pos = end_pos + 2  # bỏ qua )E

    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(''.join(out_parts))

    print(f"✅ Đã giải mã xong! Kết quả được lưu vào: {output_path}")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(f"Cách dùng: python {sys.argv[0]} <input_file> <output_file>")
        print(f"Ví dụ:     python {sys.argv[0]} input.txt output.txt")
        sys.exit(1)
    process(sys.argv[1], sys.argv[2])

File: c++_version/compiler/solve_text_txt.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <algorithm>
#include <cctype>
#include <locale>

using namespace std;

map<string, string> base_map;
map<char, map<string, string>> diacritic_map;
map<string, map<string, string>> tone_map;
map<string, string> t_map; // t0-t7: toán tử đơn

void initialize_maps() {
    // Ký tự a-z: s1→a ... s26→z
    for (int i = 1; i <= 26; ++i) {
        base_map["s" + to_string(i)] = string(1, 'a' + i - 1);
    }

    // Ký tự đặc biệt
    base_map["s27"] = ","; // dấu phẩy
    base_map["s28"] = "."; // dấu chấm

    // Chữ số: c0→0 ... c9→9
    for (int i = 0; i <= 9; ++i) {
        base_map["c" + to_string(i)] = to_string(i);
    }

    // s0 → dấu cách
    base_map["s0"] = " ";

    // Dấu phụ (diacritics)
    diacritic_map['a']["a1"] = "ă";
    diacritic_map['a']["a2"] = "â";
    diacritic_map['e']["a2"] = "ê";
    diacritic_map['o']["a2"] = "ô";
    diacritic_map['o']["a3"] = "ơ";
    diacritic_map['u']["a3"] = "ư";

    // Thanh điệu
    auto add_tones = [&](const string& base, const string& S, const string& H,
                         const string& O, const string& N, const string& A) {
        tone_map[base]["a4"] = S; // sắc
        tone_map[base]["a5"] = H; // huyền
        tone_map[base]["a6"] = O; // hỏi
        tone_map[base]["a7"] = N; // ngã
        tone_map[base]["a8"] = A; // nặng
    };

    add_tones("a", "á", "à", "ả", "ã", "ạ"); add_tones("A", "Á", "À", "Ả", "Ã", "Ạ");
    add_tones("ă", "ắ", "ằ", "ẳ", "ẵ", "ặ"); add_tones("Ă", "Ắ", "Ằ", "Ẳ", "Ẵ", "Ặ");
    add_tones("â", "ấ", "ầ", "ẩ", "ẫ", "ậ"); add_tones("Â", "Ấ", "Ầ", "Ẩ", "Ẫ", "Ậ");
    add_tones("e", "é", "è", "ẻ", "ẽ", "ẹ"); add_tones("E", "É", "È", "Ẻ", "Ẽ", "Ẹ");
    add_tones("ê", "ế", "ề", "ể", "ễ", "ệ"); add_tones("Ê", "Ế", "Ề", "Ể", "Ễ", "Ệ");
    add_tones("i", "í", "ì", "ỉ", "ĩ", "ị"); add_tones("I", "Í", "Ì", "Ỉ", "Ĩ", "Ị");
    add_tones("o", "ó", "ò", "ỏ", "õ", "ọ"); add_tones("O", "Ó", "Ò", "Ỏ", "Õ", "Ọ");
    add_tones("ô", "ố", "ồ", "ổ", "ỗ", "ộ"); add_tones("Ô", "Ố", "Ồ", "Ổ", "Ỗ", "Ộ");
    add_tones("ơ", "ớ", "ờ", "ở", "ỡ", "ợ"); add_tones("Ơ", "Ớ", "Ờ", "Ở", "Ỡ", "Ợ");
    add_tones("u", "ú", "ù", "ủ", "ũ", "ụ"); add_tones("U", "Ú", "Ù", "Ủ", "Ũ", "Ụ");
    add_tones("ư", "ứ", "ừ", "ử", "ữ", "ự"); add_tones("Ư", "Ứ", "Ừ", "Ử", "Ữ", "Ự");
    add_tones("y", "ý", "ỳ", "ỷ", "ỹ", "ỵ"); add_tones("Y", "Ý", "Ỳ", "Ỷ", "Ỹ", "Ỵ");
    // Toán tử: t0–t7
    t_map["t0"] = "=";
    t_map["t1"] = "+";
    t_map["t2"] = "-";
    t_map["t3"] = "*";
    t_map["t4"] = "/";
    t_map["t5"] = "^";
    t_map["t6"] = "//";
    t_map["t7"] = "%";
    // t8 xử lý riêng (bọc ngoặc)

}

// Giải mã chuỗi escape trong oth[...]
string decode_escape_sequence(const string& str) {
    string result;
    for (size_t i = 0; i < str.length(); ++i) {
        if (str[i] == '*' && i + 1 < str.length()) {
            // *tab → \t
            if (i + 3 < str.length() && str.substr(i+1, 3) == "tab") {
                result += '\t';
                i += 3;
            // *\n → \n
            } else if (str[i+1] == '\\' && i + 2 < str.length() && str[i+2] == 'n') {
                result += '\n';
                i += 2;
            } else {
                i++;
                if (str[i] == 'n') {
                    result += '\n';
                } else if (str[i] == 't') {
                    result += '\t';
                } else {
                    result += str[i];
                }
            }
        } else if (str[i] == '\\' && i + 1 < str.length()) {
            i++;
            if (str[i] == 'n') result += '\n';
            else if (str[i] == 't') result += '\t';
            else { result += '\\'; result += str[i]; }
        } else {
            result += str[i];
        }
    }
    return result;
}

// Giải mã một token dạng sX[modifiers] hoặc cX[modifiers]
string solve_token(const string& token) {
    size_t bracket_pos = token.find('[');
    string base_code = token.substr(0, bracket_pos);
    string decoded_char;

    if (base_map.count(base_code)) {
        decoded_char = base_map[base_code];
    } else {
        return "";
    }

    if (bracket_pos == string::npos) {
        return decoded_char;
    }

    string mods_str = token.substr(bracket_pos + 1, token.length() - bracket_pos - 2);
    stringstream mods_ss(mods_str);
    string mod;
    vector<string> modifiers;
    while (mods_ss >> mod) {
        modifiers.push_back(mod);
    }

    bool make_uppercase = false;
    string diacritic_mod = "";
    string tone_mod = "";
    int repeat_count = 1;

    for (const auto& m : modifiers) {
        if (m == "cl") {
            make_uppercase = true;
        } else if (m == "a0") {
            // đ / Đ (chỉ áp dụng với s4 = d)
            if (base_code == "s4") {
                return make_uppercase ? "Đ" : "đ";
            }
        } else if (m == "a1" || m == "a2" || m == "a3") {
            diacritic_mod = m;
        } else if (m >= "a4" && m <= "a8") {
            tone_mod = m;
        } else if (!m.empty() && isdigit(m[0])) {
            try { repeat_count = stoi(m); } catch (...) {}
        }
    }

    // Viết hoa ký tự đơn
    if (make_uppercase && decoded_char.length() == 1) {
        decoded_char[0] = toupper(decoded_char[0]);
    }

    // Dấu phụ
    if (!diacritic_mod.empty() && decoded_char.length() == 1) {
        char base_vowel = tolower(decoded_char[0]);
        if (diacritic_map.count(base_vowel) && diacritic_map.at(base_vowel).count(diacritic_mod)) {
            decoded_char = diacritic_map.at(base_vowel).at(diacritic_mod);
            if (make_uppercase) {
                map<string,string> upper_diac = {
                    {"ă","Ă"},{"â","Â"},{"ê","Ê"},{"ô","Ô"},{"ơ","Ơ"},{"ư","Ư"}
                };
                if (upper_diac.count(decoded_char)) decoded_char = upper_diac[decoded_char];
            }
        }
    }

    // Thanh điệu
    if (!tone_mod.empty()) {
        if (tone_map.count(decoded_char) && tone_map.at(decoded_char).count(tone_mod)) {
            decoded_char = tone_map.at(decoded_char).at(tone_mod);
        }
    }

    // Lặp lại
    string result;
    for (int i = 0; i < repeat_count; ++i) {
        result += decoded_char;
    }
    return result;
}

// Tách token từ chuỗi bên trong S(...)E, hỗ trợ dấu phân cách: space, comma, tab
// Ngoài ra xử lý ngoặc đơn bọc như (oth[*tab],s19,...)
vector<string> tokenize_input(const string& input) {
    vector<string> tokens;
    string current_token;
    bool in_bracket = false;  // trong []
    bool in_paren = false;    // trong ()

    for (size_t i = 0; i < input.length(); ++i) {
        char c = input[i];

        if (c == '[') {
            in_bracket = true;
            current_token += c;
        } else if (c == ']') {
            in_bracket = false;
            current_token += c;
            tokens.push_back(current_token);
            current_token.clear();
        } else if (c == '(' && !in_bracket) {
            // Bắt đầu nhóm (): bỏ qua dấu ngoặc, tiếp tục tokenize bên trong
            in_paren = true;
        } else if (c == ')' && !in_bracket) {
            // Kết thúc nhóm
            if (!current_token.empty()) {
                tokens.push_back(current_token);
                current_token.clear();
            }
            in_paren = false;
        } else if ((c == ' ' || c == ',' || c == '\t') && !in_bracket) {
            if (!current_token.empty()) {
                tokens.push_back(current_token);
                current_token.clear();
            }
        } else {
            current_token += c;
        }
    }

    if (!current_token.empty()) {
        tokens.push_back(current_token);
    }

    return tokens;
}

// Trích xuất nội dung bên trong S(...)E từ một dòng
// Trả về vector các block S(...)E tìm được trong dòng
vector<string> extract_SE_blocks(const string& line) {
    vector<string> blocks;
    size_t pos = 0;
    while (pos < line.length()) {
        size_t s_pos = line.find("S(", pos);
        if (s_pos == string::npos) break;
        // Tìm )E tương ứng — cần đếm ngoặc để tránh nhầm
        int depth = 0;
        size_t end_pos = string::npos;
        for (size_t i = s_pos + 2; i < line.length(); ++i) {
            if (line[i] == '(') depth++;
            else if (line[i] == ')') {
                if (depth == 0) {
                    // Kiểm tra xem sau ) có 'E' không
                    if (i + 1 < line.length() && line[i+1] == 'E') {
                        end_pos = i;
                        break;
                    } else {
                        // ) bình thường, không phải )E
                    }
                } else {
                    depth--;
                }
            }
        }
        if (end_pos == string::npos) break;
        string inner = line.substr(s_pos + 2, end_pos - s_pos - 2);
        blocks.push_back(inner);
        pos = end_pos + 2; // bỏ qua )E
    }
    return blocks;
}

// Xử lý và giải mã một block nội dung (đã trích từ S(...)E)
string decode_block(const string& content) {
    vector<string> tokens = tokenize_input(content);
    string result;

    for (const auto& token : tokens) {
        if (token == "s0") {
            result += " ";
        } else if (token.rfind("oth[", 0) == 0 && token.back() == ']') {
            string inner = token.substr(4, token.length() - 5);
            result += decode_escape_sequence(inner);
        } else if (!token.empty() && token[0] == 'c' && token.find('[') == string::npos) {
            // cX không có bracket → in số
            if (token.length() > 1 && isdigit(token[1])) {
                result += token.substr(1);
            }
        } else if (!token.empty() && (token[0] == 's' || token[0] == 'c')) {
            result += solve_token(token);
        // Token toán tử: t0-t7 (đơn), t8[...] (bọc ngoặc)
        } else if (!token.empty() && token[0] == 't' && token.length() > 1 && isdigit(token[1])) {
            size_t br = token.find('[');
            string t_code = token.substr(0, br == string::npos ? token.length() : br);
            if (t_code == "t8") {
                // t8[a,b,c,...] → (nội_dung)
                if (br != string::npos && token.back() == ']') {
                    string inner = token.substr(br + 1, token.length() - br - 2);
                    result += "(" + decode_block(inner) + ")";
                }
            } else if (t_map.count(t_code)) {
                result += t_map.at(t_code);
            }
        }
        // Bỏ qua các token không nhận ra
    }

    return result;
}

int main(int argc, char* argv[]) {
    setlocale(LC_ALL, "en_US.UTF-8");

    if (argc != 3) {
        cerr << "Cách dùng: " << argv[0] << " <input_file> <output_file>" << endl;
        cerr << "Ví dụ:    " << argv[0] << " input.txt output.txt" << endl;
        return 1;
    }

    string input_filename  = argv[1];
    string output_filename = argv[2];

    ifstream input_file(input_filename);
    if (!input_file.is_open()) {
        cerr << "Lỗi: Không thể mở file " << input_filename << endl;
        return 1;
    }

    ofstream output_file(output_filename);
    if (!output_file.is_open()) {
        cerr << "Lỗi: Không thể tạo file " << output_filename << endl;
        input_file.close();
        return 1;
    }

    output_file.imbue(locale(""));
    initialize_maps();

    // Đọc toàn bộ file vào một string để hỗ trợ S(...)E trải nhiều dòng
    string full_input((istreambuf_iterator<char>(input_file)),
                       istreambuf_iterator<char>());

    size_t pos = 0;
    while (pos <= full_input.length()) {
        size_t se_start = full_input.find("S(", pos);

        // Phần text thô TRƯỚC S( → fallback xử lý line-by-line
        string raw_before = (se_start == string::npos)
                            ? full_input.substr(pos)
                            : full_input.substr(pos, se_start - pos);

        if (!raw_before.empty()) {
            stringstream ss(raw_before);
            string raw_line;
            while (getline(ss, raw_line)) {
                if (raw_line.empty()) {
                    output_file << "\n";
                    continue;
                }
                vector<string> tokens = tokenize_input(raw_line);
                for (const auto& token : tokens) {
                    if (token == "s0") {
                        output_file << " ";
                    } else if (token.rfind("oth[", 0) == 0 && token.back() == ']') {
                        string inner2 = token.substr(4, token.length() - 5);
                        output_file << decode_escape_sequence(inner2);
                    } else if (!token.empty() && token[0] == 'c' && token.find('[') == string::npos) {
                        if (token.length() > 1 && isdigit(token[1])) {
                            output_file << token.substr(1);
                        }
                    } else if (!token.empty() && (token[0] == 's' || token[0] == 'c')) {
                        output_file << solve_token(token);
                    }
                }
                output_file << "\n";
            }
        }

        if (se_start == string::npos) break;

        // Tìm )E đóng tương ứng — đếm depth ngoặc, bỏ qua nội dung trong []
        int depth = 0;
        size_t end_pos = string::npos;
        bool in_sq = false;
        for (size_t i = se_start + 2; i < full_input.length(); ++i) {
            char c = full_input[i];
            if (c == '[') { in_sq = true; continue; }
            if (c == ']') { in_sq = false; continue; }
            if (in_sq) continue;
            if (c == '(') { depth++; }
            else if (c == ')') {
                if (depth == 0) {
                    if (i + 1 < full_input.length() && full_input[i+1] == 'E') {
                        end_pos = i;
                        break;
                    }
                } else {
                    depth--;
                }
            }
        }

        if (end_pos == string::npos) break;

        // Trích nội dung trong S(...)E, thay newline → space cho tokenizer
        string inner = full_input.substr(se_start + 2, end_pos - se_start - 2);
        for (char& c : inner) {
            if (c == '\n') c = ' ';
        }

        output_file << decode_block(inner);
        // Newline sau S(...)E do người dùng tự encode bằng oth[*\n]

        pos = end_pos + 2; // bỏ qua )E
    }

    input_file.close();
    output_file.close();

    cout << "✅ Đã giải mã xong! Kết quả được lưu vào: " << output_filename << endl;
    return 0;
}